### ISSUE TITLE
[Feature request] Is it possible to load the masks from a directory rather than trying to generate masks on the fly from JSON data?

### ISSUE THREAD

**Issue by sainatarajan** [dbolya/yolact#170]
Dear @dbolya Thanks for your help the previous time. I was able to run my custom dataset on this model and produce good results. However, I have less amount of training data (500 images) and hence I have performed image augmentation on both the original images and their masks to increase the training set size. Since I don't have the labels for the new set of images, is it possible to modify the code such that the model reads images and masks from a directory instead of loading from a JSON file? I think this would benefit many people who are stuck in such cases or who do not have the information of masks as a JSON file for their dataset. Thank you.

**Comment by dbolya** [comment]
I'm not sure what exactly you mean by this, can you elaborate on what your process is? You say that you don't have labels for this new data, so what exactly are you trying to do here?

And let's say you have loose masks in a directory. How would you define categories and box information on top of that (and which masks correspond to which image)? In a JSON file? But if you already have to create a JSON file, why not just put the mask inside there to begin with?

Or do you mean you have mask polygons (such as from labelme), and want to use images as masks? Because if so, you can already do that (it's still part of the JSON file, but an alternate encoding to polygons). It's called the RLE format and can allow for an arbitrary mask in the json. To create it you can use [pycocotools](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/mask.py), and it just takes the place of the list of polygon in the JSON (i.e., `my_annotation['segmentation'] = masks.encode(my_binary_nparray)`).

**Comment by sainatarajan** [comment]
Thanks for your reply. Let's say I have these two images (original image and its masks):

![a](https://user-images.githubusercontent.com/22982857/66368494-15c40b80-e9b6-11e9-9734-404f0ccd05ae.png)
![b](https://user-images.githubusercontent.com/22982857/66368500-1bb9ec80-e9b6-11e9-9141-4c33f4f9d7ec.png)

Suppose I have around 500 images and their masks stored in a directory such as `/train/images` and `/train/masks`. To increase the data size, I perform Image Augmentation and got around 5000 images and their masks. The problem is I only have the information about labels for the first set of images (500) in a JSON file. I won't be having the information about labels for the augmented images. Is it possible to load images and their label masks from a directory without requiring the JSON file that has the information about them?

**Comment by dbolya** [comment]
The question is, how do you match the old labels to the new masks? I could see doing it two ways: have you define the mapping, or have "parents" specified for the new augmented images. In either case, you will need some kind of JSON to define the pertinent information. And with the former case, you require as much information as just duplicating the annotations in the existing JSON. The latter case doesn't seem all that useful either except for this very specific circumstance, since we already have data augmentation done on the fly so you could just implement your data augmentation method in `augmentations.py`.

So my question is, why can't you just add the augmented images and masks to the JSON? You could even create the new JSON as you're generating the augmentations, copying over the label from the mask you're augmenting.

**Comment by sainatarajan** [comment]
Thank you. I didn't have a look at the `augmentation.py` file. I will take a look at it and see how I can modify it. My only concern was, once I have done augmentation, I won't be having the coordinates values for the new set of images. For eg: suppose the coordinates of a box (in mask file) are `[(100, 100), (200, 100), (200, 200), (100, 200)]`. After performing augmentation (ex: rotation), is it possible to obtain the new set of points since the original points would have been changed?

**Comment by dbolya** [comment]
You have to compute the new points given your augmentation. Take a look at how the other augmentations do it (like random rotation, which is already there).

YOLACT can't train without boxes or class labels without changing the approach (it's certainly possible, but I don't think that's what you want), so you have to compute the augmented labels for the augmented images. Even if you loaded images from a directory, that wouldn't change the need for that.

**Comment by sainatarajan** [comment]
Thanks. I will try some methods to re-compute those points in the augmentation or else I will try to read the `augmentations.py` file and try to modify it if required to suit my needs. I will close this issue for now.
